salt/daemons/masterapi.py

```diff
@@ -187,11 +187,51 @@ def clean_old_jobs(opts):
         mminion.returners[fstr]()
 
 
+def mk_key(opts, user):
+    if salt.utils.is_windows():
+        # The username may contain '\' if it is in Windows
+        # 'DOMAIN\username' format. Fix this for the keyfile path.
+        keyfile = os.path.join(
+            opts['cachedir'], '.{0}_key'.format(user.replace('\\', '_'))
+        )
+    else:
+        keyfile = os.path.join(
+            opts['cachedir'], '.{0}_key'.format(user)
+        )
+
+    if os.path.exists(keyfile):
+        log.debug('Removing stale keyfile: {0}'.format(keyfile))
+        if salt.utils.is_windows() and not os.access(keyfile, os.W_OK):
+            # Cannot delete read-only files on Windows.
+            os.chmod(keyfile, stat.S_IRUSR | stat.S_IWUSR)
+        os.unlink(keyfile)
+
+    key = salt.crypt.Crypticle.generate_key_string()
+    cumask = os.umask(191)
+    with salt.utils.fopen(keyfile, 'w+') as fp_:
+        fp_.write(key)
+    os.umask(cumask)
```

Comment:
@cachedout here we change umask to create the key file read-only for the owner only. Then we restore our umask to return the normal behavior for other files. This looks correct for me.
Also this code wasn't touched by me, I've just moved it into a separate function.